[pytree] add access api

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117771

This PR introduces an API to use KeyPaths to actually access values on pytrees.

Differential Revision: [D52881260](https://our.internmc.facebook.com/intern/diff/D52881260/)

suo committed:
[pytree] add access api

This PR introduces an API to use KeyPaths to actually access values on pytrees.

Differential Revision: [D52881260](https://our.internmc.facebook.com/intern/diff/D52881260/)

[ghstack-poisoned]

suo committed:
Update on "[pytree] add access api"


This PR introduces an API to use KeyPaths to actually access values on pytrees.

Differential Revision: [D52881260](https://our.internmc.facebook.com/intern/diff/D52881260/)

[ghstack-poisoned]

zou3519 commented on torch/utils/_pytree.py:
```diff
@@ -1476,3 +1497,10 @@ def tree_map_with_path(
     return "".join([str(k) for k in kp])
+
+
+def key_get(obj: Any, kp: KeyPath) -> Any:
+    """Given an object and a key path, return the value at the key path."""
+    for k in kp:
+        obj = k.access(obj)
+    return obj
```
bikeshedding: the API is called "key_get", but the accessors are called "access". Can we make them  more consistent?

zou3519 replied:
ideas:
- access_with_key/access ?
- key_get/get ?

Eh, key_get / access is growing on me some more.

suo replied:
lol I hate API design

zou3519 reviewed (approved):
LGTM, but I want to bikeshed the naming a bit

suo commented:
> key_get/get

I think this is probably the best if we want them to be the same term, since `key_access` sounds silly. lol

suo committed:
Update on "[pytree] add access api"


This PR introduces an API to use KeyPaths to actually access values on pytrees.

Differential Revision: [D52881260](https://our.internmc.facebook.com/intern/diff/D52881260/)

[ghstack-poisoned]

suo commented:
@pytorchbot merge -f "merged internally"
